Fix some Gradle deprecation warnings

jonhkr opened the pull request:
Find the guide for upgrading here: https://docs.gradle.org/6.6/userguide/upgrading_version_5.html#changes_6.0

Changes made in this PR:
- Use `java-library` instead of `java` plugin (to have access to the `api` configuration)
- Use `api` instead of `compile`
- Use `testImplementation` instead of `testCompile`
- Use `testRuntimeOnly` instead of `testRuntime`

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

jonhkr committed:
Fix some Gradle deprecation warnings
- Use `implementation` instead of `compile`
- Use `testImplementation` instead of `testCompile`
- Use `testRuntimeOnly` instead of `testRuntime`

jonhkr committed:
Fix dependencies

Gradle `compile` configuration exposes the dependencies used by the
project to consumers, by replacing it with `implementation` this
behaviour changed.

The `implementation` configuration only exposes
that library to the compile and runtime classpath, but consumers
don't have access to it. As some projects used previously `compile`
exposed libraries, we need to supply those explicitly when using
the `implementation` configuration.

jonhkr commented:
The generated poms changed a bit. 
- Dependencies in the `compile` scope were moved to `runtime`.
- Some dependencies were introduced (the ones that were previously exported by the `compile` configuration.)

[poms.zip](https://github.com/apache/kafka/files/5179804/poms.zip)

jonhkr commented:
Hey @omkreddy @ijuma, would you mind taking a look into this?

ijuma commented:
Thanks for the PR.

> Dependencies in the compile scope were moved to runtime.

This seems like an issue, no?

jonhkr committed:
Use Gradle api configuration to not change the generated poms

By using the implementation configuration the generated poms changed,
libraries in the compile scope were moved to runtime.
The api configuration provided by the `java-library` plugin has
the same behaviour as the deprecated compile configuration.

jonhkr commented:
@ijuma Yes, it was.
I've updated the code to use the `api` configuration provided by the `java-library` plugin. This configuration has the same behaviour of the `compile` that was deprecated.

Now the generated poms should look exactly as before.

jonhkr commented:
@ijuma Could you take a look?

divijvaidya commented:
We are currently running gradle v8 and these warning have been fixed since this PR was opened. I am closing this PR. Please feel free to re-open if you still observe these deprecation warnings.
